src/compressor/QatAccel.cc

```diff
@@ -48,22 +53,112 @@ bool QatAccel::init(const std::string &alg) {
-  if (rc != QZ_OK && rc != QZ_DUPLICATE && rc != QZ_NO_HW)
+  if (rc != QZ_OK && rc != QZ_DUPLICATE)
     return false;
-
   rc = qzSetupSession(session.get(), &params);
-  if (rc != QZ_OK && rc != QZ_DUPLICATE && rc != QZ_NO_HW ) {
-    qzTeardownSession(session.get());
-    qzClose(session.get());
+  if (rc != QZ_OK) {
     return false;
   }
+  return true;
+}
+
+// put the session back to the session pool in a RAII manner
+struct cached_session_t {
+  cached_session_t(QatAccel* accel)
+    : accel{accel} {
+    std::unique_lock lock{accel->mutex};
+    if (!accel->sessions.empty()) {
+      session = std::move(accel->sessions.back());
+      accel->sessions.pop_back();
+      return;
+    }
 
+    // If there are no available session to use, we try allocate a new
+    // session. When we failed to allocate a new session, we block on here,
+    // waiting a available session;
+    QzSessionParams_T params = {(QzHuffmanHdr_T)0,};
+    QatAccel::session_ptr s{new struct QzSession_S()};
+    memset(s.get(), 0, sizeof(struct QzSession_S));
+    if (get_qz_params(accel->alg_name, params) && setup_session(s, params)) {
+      session = std::move(s);
+    } else {
+      accel->cond.wait(lock, [accel] { return !accel->sessions.empty(); });
+      session = std::move(accel->sessions.back());
+      accel->sessions.pop_back();
+    }
+  }
+  ~cached_session_t() {
+    std::scoped_lock lock{accel->mutex};
+    // if the cache size is still under its upper bound, the current session is put into
+    // accel->sessions. otherwise it's released right
+    uint64_t sessions_num = g_ceph_context->_conf.get_val<uint64_t>("qat_compressor_session_max_number");
+    if (accel->sessions.size() < sessions_num) {
+      accel->sessions.push_back(std::move(session));
+      accel->cond.notify_one();
+    }
+  }
+  struct QzSession_S* get() {
+    assert(static_cast<bool>(session));
+    return session.get();
+  }
+  QatAccel* accel;
+  QatAccel::session_ptr session;
+};
+
+QatAccel::QatAccel() {}
+
+QatAccel::~QatAccel() {
+  // First, we should uninitialize all QATzip session that disconnects all session
+  // from a hardware instance and deallocates buffers.
+  sessions.clear();
+  // Then we close the connection with QAT.
+  // where the value of the parameter passed to qzClose() does not matter. as long as
+  // it is not nullptr.
+  qzClose((QzSession_T*)1);
+}
+
+bool QatAccel::init(const std::string &alg) {
+  QzSessionParams_T params = {(QzHuffmanHdr_T)0,};
+  std::scoped_lock lock(mutex);
+  if (!alg_name.empty()) {
+    return true;
+  }
+
+  uint64_t sessions_num = g_ceph_context->_conf.get_val<uint64_t>("qat_compressor_session_min_number");
+  dout(20) << "Currently, sessions number is: " << sessions_num << dendl;
+  std::generate_n(
+  std::back_inserter(sessions),
+  sessions_num,
+  [] {
+    session_ptr s{new struct QzSession_S()};
+    memset(s.get(), 0, sizeof(struct QzSession_S));
+    return s;
+  });
```

Comment:
is there a need to preallocate sessions this way, now that `cached_session_t()` will allocate them on-demand?

removing this would simplify the design back down to a single config variable (the maximum size for `~cached_session_t()`)

Reply:
No, it can be removed. And I only keep `qat_compressor_session_max_number` config variable.